Safe entanglement

juliapath opened the pull request:
Hope this is better. I commented all instances of unsafePerformIO and unsafeInterleaveIO. Since the rest of the code is safe, I hope this is convincing in regards to referential transparency.

kwf commented:
Thank you! :)

Commit:
Commentate and improve readability of Observe.hs

juliapath commented:
Note that this pull request is to the safe-entanglement branch, not to master.

Commit:
Make Shape an injective type family

This allows the constraint solver to infer a ~ a' from Shape a ~ Shape a' make
some type annotations redundant. The case, that one data type is the Shape of
two different data types doesn't really seem useful, as one could just define
two newtypes of that type, as we do with GShape. Ideally we would make this
would be a data family which would allow a ~ a' and Shape ~ s to be inferred,
from Shape a ~ s a, but sadly it is not possible to define default data
families.

Commit:
Generalise unzipWith and unzipWithM

This is also a lot easier to do with Shape being injective, which is why I
decided to merge that change into this branch.

Commit:
Remove old items from todo list

plaidfinch commented on src/Test/StrictCheck/Shaped.hs:
```diff
@@ -269,16 +269,16 @@ foldM alg = alg <=< traverse (translateA @a (foldM alg)) . unwrap
+unfoldM coalg = fmap Wrap . traverse (translateA (unfoldM coalg)) <=< coalg
 
--- TODO: mapM, foldM, unfoldM, ...
+-- TODO: mapM, ...
```
`mapM` is called "translateA` here. I think I should rename it...

juliapath replied:
I would have expected `mapM` to have a type similar to `(forall x. f x -> m (g x)) -> f % a -> m (g % a)`, similar to `foldM` and `unzipWithM`. But that can already be archived rather easily with `foldM` and `unfoldM`. `foldM` requires `f` to be traversable and `unfoldM` requires `g` to be traversable, so they both versions are useful. So I think the name `translateA` is better than `mapM`. It is also not taken by `Prelude`.

plaidfinch commented on src/Test/StrictCheck/Shaped.hs:
```diff
@@ -126,7 +126,7 @@ class Typeable a => Shaped (a :: *) where
   -- structure in an @a@, parameterized by the functor @f@, which is wrapped
   -- around any fields (of any type) in the original @a@.
-  type Shape a :: (* -> *) -> *
+  type Shape a = (result :: (* -> *) -> *) | result -> a
```
I like this, especially the way it aids type inference. I think you're right that it's unlikely to be an issue in practice, given that it's rather unlikely for an existing datatype to even have the correct *kind* to be a `Shape`.

plaidfinch commented on src/Test/StrictCheck/Observe.hs:
```diff
@@ -173,34 +183,42 @@ entangle a = do
-          $ flatA)
+    -- The to be entangled value with all its recursive children entangled. We
+    -- still need to entangle the value itself.
+    entangledChildren :: IO (a, IO (Shape a Demand))
```
Instead of using `match` directly, I think there's a way to express this using `translateA`, which is the `traverse` equivalent for `Shape`s. I'll check this out later.

juliapath replied:
Good idea. Implemented it on my branch.

Commit:
A better generalisation of unzipWith and unzipWithM

For reference, here are the two types, of the split function:

Old:
(forall x sx. sx ~ Shape x (Product ((%) h) ((%) g))  => f sx -> (g sx, h sx))

New:
(forall x sx. sx ~ (Shape x ((%) g), Shape x ((%) h)) => f sx -> (g sx, h sx))

The new type unzips the argument first and calls split then, thus giving it easy
access to the two branches.

There is a slightly more general version, but it is less readable and not more
general in a useful way:
(forall x. f (Shape x ((%) g), Shape x ((%) h))
     -> (g ((Shape x ((%) g)), t1), h (t2, (Shape x ((%) h)))))

Commit:
Use translateA in entangleShape

plaidfinch commented on src/Test/StrictCheck/Shaped.hs:
```diff
@@ -313,32 +313,34 @@ interleave p = unfold (fmap (project p)) . p
 -- @h % a@.
```
I'm not sure I follow this generalization of `unzipWith`; can you elaborate on what's going on here? It seems strange to have the `(%)` type occur in the splitting function; the rest of these recursion schemes have a non-recursive operation as their argument, then extend that to the recursive (interleaved) case. Whereas this type indicates that the `split` function could deconstruct the results of the unzipping, kind of like a histomorphism. Is that what's intended here (and if so, what's the intended use)?

juliapath replied:
Well the former type of the splitting function was `forall x . f x -> (g x, h x)`, but actually we know a bit more about `x`. I just put a typed hole were `split` was and took that as the new type of `split`. Now `split` has access to the already unzipped structure under the constructor that it splits right now. In my second commit I just changed the order of `split` and `crunch` (changing crunch a bit), as that yields a more useful type for split.

I originally did this, because I thought I could use the more general `zipWithM` to implement `entangleShape`, but it turned out not useful to be useful for that. Still it is strictly more general and the behaviour is the same.

plaidfinch commented on src/Test/StrictCheck/Shaped.hs:
```diff
@@ -313,32 +313,34 @@ interleave p = unfold (fmap (project p)) . p
+                       . fmap crunch)
 
 -- Some helpers for zipping and unzipping...
 
```
I'm not sure `crunch` is the right name for this operation any more (we've changed what it does). I can't come up with a good verb for it, though...

juliapath replied:
I didn't really understand the name in the first place, so I just left it.

juliapath replied:
`unzipShape`?
